source/extensions/common/wasm/context.cc

```diff
@@ -914,6 +914,8 @@ WasmResult Context::httpCall(absl::string_view cluster, const Pairs& request_hea
     token = next_http_call_token_++;
   }
   auto& handler = http_request_[token];
+  handler.context_ = this;
```

Comment:
What is the case where this could cause a segfault?  If  the send() fails, then it will erase the handler.  If it succeeds then it will set those values and  If this is a reentrant call from within send() we have an issue to remove those.

Reply:
I created a reproducer at https://github.com/robertpanzer/http-call-crash
When the DNS name configured for the cluster cannot be resolved it seems like the handler is called synchronously from send() before `handler.context_` is set.
The stack trace in #472 seems to show that too:
```
[2020-04-01 11:42:40.135][18205][critical][backtrace] [bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:104] Caught Segmentation fault, suspect faulting address 0x158
[2020-04-01 11:42:40.135][18205][critical][backtrace] [bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:91] Backtrace (use tools/stack_decode.py to get line numbers):
[2020-04-01 11:42:40.135][18205][critical][backtrace] [bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:92] Envoy version: f6d0282f7ab42dfd215ae59112edbab3e70f2d3b/1.14.0-dev/Modified/DEBUG/BoringSSL
[2020-04-01 11:42:40.299][18205][critical][backtrace] [bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #0: Envoy::SignalAction::sigHandler() [0x55c07362340c]
[2020-04-01 11:42:40.299][18205][critical][backtrace] [bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #1: __restore_rt [0x7f147f1a5890]
[2020-04-01 11:42:40.416][18205][critical][backtrace] [bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #2: Envoy::Extensions::Common::Wasm::Context::AsyncClientHandler::onSuccess() [0x55c0717bba65]
[2020-04-01 11:42:40.479][18205][critical][backtrace] [bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #3: Envoy::Http::AsyncRequestImpl::onComplete() [0x55c07305df4a]
[2020-04-01 11:42:40.517][18205][critical][backtrace] [bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #4: Envoy::Http::AsyncStreamImpl::closeLocal() [0x55c07305d073]
[2020-04-01 11:42:40.554][18205][critical][backtrace] [bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #5: Envoy::Http::AsyncStreamImpl::encodeData() [0x55c07305d2e4]
[2020-04-01 11:42:40.589][18205][critical][backtrace] [bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #6: Envoy::Http::AsyncStreamImpl::sendLocalReply()::{lambda()#1}::operator()() [0x55c073062d5e]
[2020-04-01 11:42:40.624][18205][critical][backtrace] [bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #7: std::_Function_handler<>::_M_invoke() [0x55c073062bfc]
[2020-04-01 11:42:40.661][18205][critical][backtrace] [bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #8: std::function<>::operator()() [0x55c070ac9c22]
[2020-04-01 11:42:40.697][18205][critical][backtrace] [bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #9: Envoy::Http::Utility::sendLocalReply() [0x55c073306795]
[2020-04-01 11:42:40.733][18205][critical][backtrace] [bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #10: Envoy::Http::AsyncStreamImpl::sendLocalReply() [0x55c073060529]
[2020-04-01 11:42:40.766][18205][critical][backtrace] [bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #11: Envoy::Router::Filter::sendNoHealthyUpstreamResponse() [0x55c07306f1ad]
[2020-04-01 11:42:40.879][18205][critical][backtrace] [bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #12: Envoy::Router::Filter::decodeHeaders() [0x55c07306de72]
[2020-04-01 11:42:40.968][18205][critical][backtrace] [bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #13: Envoy::Http::AsyncStreamImpl::sendHeaders() [0x55c07305d731]
[2020-04-01 11:42:41.059][18205][critical][backtrace] [bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #14: Envoy::Http::AsyncRequestImpl::initialize() [0x55c07305c352]
[2020-04-01 11:42:41.160][18205][critical][backtrace] [bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #15: Envoy::Http::AsyncClientImpl::send() [0x55c07305bbd6]
[2020-04-01 11:42:41.279][18205][critical][backtrace] [bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #16: Envoy::Extensions::Common::Wasm::Context::httpCall() [0x55c07178918e]
[2020-04-01 11:42:41.314][18205][critical][backtrace] [bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #17: Envoy::Extensions::Common::Wasm::Exports::http_call() [0x55c0717cda44]
[2020-04-01 11:42:41.350][18205][critical][backtrace] [bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #18: Envoy::Extensions::Common::Wasm::ConvertFunctionWordToUint32<>::convertFunctionWordToUint32() [0x55c0717ed3cf]
[2020-04-01 11:42:41.387][18205][critical][backtrace] [bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #19: absl::base_internal::Callable::Invoke<>() [0x55c07183dced]
[2020-04-01 11:42:41.422][18205][critical][backtrace] [bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #20: absl::base_internal::Invoke<>() [0x55c07183d9da]
[2020-04-01 11:42:41.461][18205][critical][backtrace] [bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #21: absl::utility_internal::apply_helper<>() [0x55c07183d831]
[2020-04-01 11:42:41.493][18205][critical][backtrace] [bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #22: _ZN4absl5applyIRPFjPvjjjjjjjjjjERSt5tupleIJPN5Envoy10Extensions6Common4Wasm7ContextEjjjjjjjjjjEEEEDTclsr16utility_internalE12apply_helperclsr4abslE7forwardIT_Efp_Eclsr4abslE7forwardIT0_Efp0_EtlNS_16utility_internal3GenImXsr3std10tuple_sizeINSt16remove_referenceISF_E4typeEEE5valueEE4typeEEEEOSE_OSF_ [0x55c07183ba72]
[2020-04-01 11:42:41.525][18205][critical][backtrace] [bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #23: Envoy::Extensions::Common::Wasm::V8::V8::registerHostFunctionImpl<>()::{lambda()#1}::operator()() [0x55c07183b8cf]
[2020-04-01 11:42:41.563][18205][critical][backtrace] [bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #24: Envoy::Extensions::Common::Wasm::V8::V8::registerHostFunctionImpl<>()::{lambda()#1}::__invoke() [0x55c07183b753]
[2020-04-01 11:42:41.603][18205][critical][backtrace] [bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #25: wasm::FuncData::v8_callback() [0x55c07185908f]
```

Reply:
Generally we want to avoid reentrant calls into the WasmVM.   Here is a sketch of the solution using that model: https://github.com/jplevyak/envoy-wasm/pull/new/http-reentrant

I need to flesh it out for the other cases (onFailure, grpc).

Reply:
I cannot comment on your PR so I’m adding it here:
Doesn’t your PR still use the token which should also not be initialized in that case. 
Of course it would not result in a segfault as it’s just an int, but I guess it would result in some other unknown behavior?